Brush up modbus docs (#3478)
OttoWinter/esphomedocs: components/modbus_controller.rst

The transceiver connects to the UART of the MCU. For ESP32, pin ``16`` to ``TXD`` and pin ``17`` to ``RXD`` are the default ones but any other pins can be used as well. ``3.3V`` to ``VCC`` and naturally ``GND`` to ``GND``.

> On the bus side, you need 120 Ohm termination resistors at the ends of the bus cable as per Modbus standard. Some transceivers have this already solderes onboard, and some slave devices may have them preinstalled with a jumper or a dip switch.

Reviewer: ```suggestion
On the bus side, you need 120 Ohm termination resistors at the ends of the bus cable as per Modbus standard. Some transceivers have this already soldered onboard, while some slave devices may have them available via a jumper or a DIP switch.
```

.. note::

    If you are using an ESP8266, serial logging may cause problems reading from UART. For best results, hardware serial is recommended. Software serial may not be able to read all received data if other components spend a lot of time in the ``loop()``.

    For hardware serial only a limited set of pins can be used. Either ``tx_pin: GPIO1`` and ``rx_pin: GPIO3``  or ``tx_pin: GPIO15`` and ``rx_pin: GPIO13``.

>     The disadvantage of using the hardware uart is that you can't use serial logging because the serial logs would be sent to the Modbus device and cause errors.

Reviewer: ```suggestion
    The disadvantage of using the hardware UART is that you can't use serial logging because the serial logs would be sent to the Modbus device(s) instead, causing errors.
```

- **modbus_id** (*Optional*, :ref:`config-id`): Manually specify the ID of the ``modbus`` hub.

- **address** (**Required**, :ref:`config-id`): The Modbus address of the slave device

- **command_throttle** (*Optional*, :ref:`config-time`): minimum time in between 2 requests to the device. Default is ``0ms``.

>   Some Modbus slave devices limit the rate of requests from the master, the interval between sending requests can be altered.

Reviewer: ```suggestion
  Some Modbus slave devices limit the rate of requests from the master, so this allows the interval between requests to be altered.
```

  responds to a command, it'll be marked online again.

Example
-------
The following code creates a ``modbus_controller`` hub talking to a ModBUS device at address ``1`` with ``115200`` bps

ModBUS sensors can be directly defined (inline) under the ``modbus_controller`` hub or as standalone components
Technically there is no difference between the "inline" and the standard definitions approach.

> 

Reviewer 2: Why remove the example header here?

Author: Added back. Thanks

      unit_of_measurement: kWh
      accuracy_decimals: 1

    - platform: modbus_controller
      modbus_controller_id: sdm
      name: "Total reactive energy"
      #   address: 0x158
      #   register_type: "read"
      custom_command: [0x2, 0x4, 0x1, 0x58, 0x00, 0x02]
      ## the command returns an float value using 4 bytes
      lambda: |-
          ESP_LOGD("Modbus Sensor Lambda","Got new data" );
          union {
            float float_value;
            uint32_t raw;
          } raw_to_float;
          if (data.size() < 4 ) {
            ESP_LOGE("Modbus Sensor Lambda", "invalid data size %d",data.size());
            return NAN;
          }
          raw_to_float.raw =   data[0] << 24 | data[1] << 16 | data[2] << 8 |  data[3];
          ESP_LOGD("Modbus Sensor Lambda", "FP32 = 0x%08X => %f", raw_to_float.raw, raw_to_float.float_value);
          return raw_to_float.float_value;
        unit_of_measurement: kVArh
        accuracy_decimals: 1

.. _modbus_register_count:

Optimizing modbus communications
--------------------------------

``register_count`` is an option only required for uncommon response encodings or to optimizie modbus communications.

> It describes the number of registers this data point spans. Overrides the defaults determined by ``value_type``. If no value for ``register_count`` is provided, it is calculated based on the register type. The default size for 1 register is 16 bits (1 Word). Some devices are not adhering to this convention and have registers larger than 16 bits. In this case ``register_count`` and  ``response_size`` must be set. For example, if your modbus device uses 1 registers for a FP32 value instead the default of two set ``register_count: 1`` and ``response_size: 4``.

Reviewer: ```suggestion
It describes the number of registers this data point spans, overriding the defaults determined by ``value_type``. If no value for ``register_count`` is provided, it is calculated based on the register type. The default size for one register is 16 bits (one word). Some devices are not adhering to this convention and have registers larger than 16 bits. In this case, ``register_count`` and  ``response_size`` must be set. For example, if your Modbus device uses one register for a FP32 value (instead of the default of two), set ``register_count: 1`` and ``response_size: 4``.
```

``register_count`` can also be used to skip a number of registers in consecutive range.

An example is an SDM meter, with interesting data in register addresses 0, 2, 4 and 6:

.. code-block:: yaml

    - platform: modbus_controller
      name: "Voltage Phase 1"
      address: 0
      register_type: "read"
      value_type: FP32

    - platform: modbus_controller
      name: "Voltage Phase 2"
      address: 2
      register_type: "read"
      value_type: FP32

    - platform: modbus_controller
      name: "Voltage Phase 3"
      address: 4
      register_type: "read"
      value_type: FP32

    - platform: modbus_controller
      name: "Current Phase 1"
      address: 6
      register_type: "read"
      value_type: FP32
      accuracy_decimals: 1

The configuration above will generate *one* modbus command *read multiple registers from 0 to 6*.

Maybe you don’t care about the data in register addresses 2 and 4, which are voltage values for Phase 2 and Phase 3 (or you have a SDM-120).

> Of course, you can delete the sensors your don’t care about. But then you have a gap in the addresses. If you remove the registers at address 2 and 4 then *two* commands will be generated *read register 0* and *read register 6*. To avoid the generation of multiple commands and reduce the amount of uart communication ``register_count`` can be used to fill the gaps:

Reviewer: ```suggestion
Of course, you can delete the sensors your don’t care about, but then you'd have a gap in the addresses. If you remove the registers at address 2 and 4, *two* commands will be generated -- *read register 0* and *read register 6*. To avoid generating multiple commands and thus reduce activity on the bus, ``register_count`` can be used to fill the gaps:
```

.. code-block:: yaml

    - platform: modbus_controller
      name: "Voltage Phase 1"
      address: 0
      unit_of_measurement: "V"
      register_type: "read"
      value_type: FP32
      register_count: 6

    - platform: modbus_controller
      name: "Current Phase 1"
      address: 6
      register_type: "read"
      value_type: FP32

> Because the option `register_count: 6` is used for the first sensor, *one* command *read multiple registers from 0 to 6* will be used but the values in between will be ignored.

Reviewer: ```suggestion
Because the option ``register_count: 6`` is used for the first sensor, *one* command *read multiple registers from 0 to 6* will be used but the values in between will be ignored.
```

      id: array_rated_power
      name: "array_rated_power"
      address: 0x3002
      unit_of_measurement: "W"
      register_type: read
      value_type: U_DWORD_R
      accuracy_decimals: 1
      filters:
        - multiply: 0.01

    - platform: modbus_controller
      modbus_controller_id: epever
      id: battery_rated_voltage
      name: "battery_rated_voltage"
      address: 0x3004
      unit_of_measurement: "V"
      register_type: read
      value_type: U_WORD
      accuracy_decimals: 1
      filters:
        - multiply: 0.01

    - platform: modbus_controller
      modbus_controller_id: epever
      id: battery_rated_current
      name: "battery_rated_current"
      address: 0x3005
      unit_of_measurement: "A"
      register_type: read
      value_type: U_WORD
      accuracy_decimals: 1
      filters:
        - multiply: 0.01

    - platform: modbus_controller
      modbus_controller_id: epever
      id: battery_rated_power
      name: "battery_rated_power"
      address: 0x3006
      unit_of_measurement: "W"
      register_type: read
      value_type: U_DWORD_R
      accuracy_decimals: 1
      filters:
        - multiply: 0.01

    - platform: modbus_controller

>       modbus_controller_id: epever id: charging_mode

Reviewer 2: Looks like something went wrong on this line?

Author: Not anymore

.. note::

>     Write support is only implemented for switches and selects.

Reviewer: ```suggestion
    Write support is only implemented for switches and selects; however, the C++ code provides the required API to write to a Modbus device.
```

>     However the C++ code provides the required API to write to a Modbus device.

Reviewer: (See above)
```suggestion
```
